O2p authentication

jackwrfuller opened the pull request.

shreddedbacon commented:
Just linking here for my own benefit for :eyes: 
* https://github.com/jackwrfuller/build-deploy-tool/pull/1
* https://github.com/jackwrfuller/lagoon-charts/pull/1
* https://github.com/jackwrfuller/lagoon-go-client/pull/1
* https://github.com/oauth2-proxy/oauth2-proxy/compare/master...jackwrfuller:oauth2-proxy:master

shreddedbacon commented on services/keycloak/startup-scripts/00-configure-lagoon.sh:
```diff
@@ -677,6 +680,32 @@ function add_lagoon-cli_client {
+            "serviceAccountsEnabled": false,
+            "standardFlowEnabled": true,
+            "directAccessGrantsEnabled": true,
+            "redirectUris": ["'"$REDIRECT_URI"'"],
```
This doesn't need a redirecturi to be set since it requires the clientid and secret, which would mean that `OAUTH2_PROXY_FRONTEND_URL` is not requried to be passed to the [keycloak deployment here](https://github.com/jackwrfuller/lagoon-charts/pull/1/files#diff-59729e08f9b738e4e33af5c90bab40e36de7ab06615e1cc017f9996dc2a8f36bR52)

jackwrfuller replied:
removed the specific URI

shreddedbacon commented on Makefile:
```diff
@@ -804,6 +806,7 @@ endif
 		USE_CALICO_CNI=false \
 		LAGOON_SSH_PORTAL_LOADBALANCER=$(LAGOON_SSH_PORTAL_LOADBALANCER) \
 		LAGOON_FEATURE_FLAG_DEFAULT_ROOTLESS_WORKLOAD=enabled \
+		LAGOON_FEATURE_FLAG_OAUTH2PROXY_DOMAIN="http://lagoon-oauth2proxy.$$($(KUBECTL) -n ingress-nginx get services ingress-nginx-controller -o jsonpath='{.status.loadBalancer.ingress[0].ip}').nip.io"
```
```suggestion
		LAGOON_FEATURE_FLAG_OAUTH2PROXY_DOMAIN="http://lagoon-oauth2proxy.$$($(KUBECTL) -n ingress-nginx get services ingress-nginx-controller -o jsonpath='{.status.loadBalancer.ingress[0].ip}').nip.io" \
```

shreddedbacon reviewed (changes requested):
I can't actually get this to work properly, there are things wrong with the charts too which i've left comments on

shreddedbacon commented:
For ease of testing, you should update the makefile to point to your charts fork:

```
CHARTS_TREEISH = o2p-no-sidecar
CHARTS_REPOSITORY = https://github.com/jackwrfuller/lagoon-charts.git
```

jackwrfuller commented:
> For ease of testing, you should update the makefile to point to your charts fork:
> 
> ```
> CHARTS_TREEISH = o2p-no-sidecar
> CHARTS_REPOSITORY = https://github.com/jackwrfuller/lagoon-charts.git
> ```

Done, i usually omit it from PRs so that the PR is immediately mergable.

shreddedbacon commented on services/keycloak/startup-scripts/00-configure-lagoon.sh:
```diff
@@ -677,6 +680,31 @@ function add_lagoon-cli_client {
     echo '{"protocol":"openid-connect","config":{"id.token.claim":"true","access.token.claim":"true","userinfo.token.claim":"true","user.attribute":"lagoon-uid","claim.name":"lagoon.user_id","jsonType.label":"int","multivalued":""},"name":"Lagoon User ID","protocolMapper":"oidc-usermodel-attribute-mapper"}' | /opt/keycloak/bin/kcadm.sh create -r ${KEYCLOAK_REALM:-master} clients/$CLIENT_ID/protoc ...
 }
 
+function add_lagoon-oauth2proxy_client {
```
You also need to add the client creation to the realm import script, see this commit as an example https://github.com/uselagoon/lagoon/commit/8bff110ee114fc143dfc5236c3ba57b920b98bfc#diff-6cb8ba885515e077cce12118670e86727fef2319c6d0088ef5432428e11a84e2

shreddedbacon replied:
We update both files, so that eventually we can trim older migrations/creations from this `00-configure-lagoon.sh` script to make keycloak start up faster.

Adding it to the import json file means that for fresh installs, everything will be imported from the start and the `00-configure-lagoon.sh` script essentially runs mostly `no-op`

jackwrfuller replied:
@shreddedbacon I've added the client to the realm import. I've copied over the config from the other script, which i assume is sufficient for the realm import as well?

shreddedbacon commented on Makefile:
```diff
@@ -938,6 +941,14 @@ k3d/push-local-build-image:
 		&& docker tag lagoon/build-deploy-image:local $$IMAGE_REGISTRY/build-deploy-image:$(BUILD_DEPLOY_IMAGE_TAG) \
 		&& docker push $$IMAGE_REGISTRY/build-deploy-image:$(BUILD_DEPLOY_IMAGE_TAG)
 
+.PHONY: k3d/push-local-o2p-image
+k3d/push-local-o2p-image:
+	@export KUBECONFIG="$$(pwd)/kubeconfig.k3d.$(CI_BUILD_TAG)" && \
+		export IMAGE_REGISTRY="registry.$$($(KUBECTL) -n ingress-nginx get services ingress-nginx-controller -o jsonpath='{.status.loadBalancer.ingress[0].ip}').nip.io/library" \
+		&& docker login -u admin -p Harbor12345 $$IMAGE_REGISTRY \
+		&& docker tag quay.io/oauth2-proxy/oauth2-proxy:latest $$IMAGE_REGISTRY/oauth2-proxy:o2p-authentication \
+		&& docker push $$IMAGE_REGISTRY/oauth2-proxy:o2p-authentication
```
What is the purpose of this target? It accomplishes the same thing that `make k3d/push-images IMAGES=oauth2-proxy` does no?

jackwrfuller replied:
ah don't think so? I forked o2p and made my changes to the fork, then built the images locally, then imported the image with the above target, like how we do it BDT. This was more to help me test my custom o2p image, not sure if this would be included in the end or nah

shreddedbacon replied:
Yeah this will need to be removed once all the changes are put here

shreddedbacon reviewed (commented):
We should also have a chat about a new permission, rather than rely on `project:view` in the future.

Since we have `environment:ssh:development` and `environment:ssh:production` ([see](https://github.com/uselagoon/lagoon/blob/v2.28.0/services/keycloak/lagoon-realm-base-import.json#L2047-L2048)), it might make sense to have something like this 

Two new `resource:scopes` like
* `environment:viewRoute:development`
* `environment:viewRoute:production`

Probably configured the same way on the same roles that `ssh` are currently configured on.

This way, in the future if custom roles are added to lagoon, it would be possible to create users that will only be able to `viewRoute` and have 0 access to anything else in lagoon (not even viewing the project in the UI, etc)

jackwrfuller commented:
@shreddedbacon I've updated the PRs to now use the new API call. I have also updated ur previous comment to include links to my machinery PR and also my o2p fork.

shreddedbacon commented on services/keycloak/startup-scripts/00-configure-lagoon.sh:
```diff
@@ -408,7 +408,7 @@ function remove_deleteall_permissions_scopes {
 
   ENVIRONMENT_RESOURCE_ID=$(/opt/keycloak/bin/kcadm.sh get -r lagoon clients/$CLIENT_ID/authz/resource-server/resource?name=environment --config $CONFIG_PATH | jq -r '.[0]["_id"]')
-  /opt/keycloak/bin/kcadm.sh update clients/$CLIENT_ID/authz/resource-server/resource/$ENVIRONMENT_RESOURCE_ID --config $CONFIG_PATH -r ${KEYCLOAK_REALM:-master} -s 'scopes=[{"name":"deploy:production"},{"name":"addOrUpdate:production"},{"name":"viewAll"},{"name":"storage"},{"name":"addOrUpdate:development"},{"name":"update:development"},{"name":"ssh:development"},{"name":"delete:development"},{" ...
+  /opt/keycloak/bin/kcadm.sh update clients/$CLIENT_ID/authz/resource-server/resource/$ENVIRONMENT_RESOURCE_ID --config $CONFIG_PATH -r ${KEYCLOAK_REALM:-master} -s 'scopes=[{"name":"deploy:production"},{"name":"addOrUpdate:production"},{"name":"viewAll"},{"name":"storage"},{"name":"addOrUpdate:development"},{"name":"update:development"},{"name":"ssh:development"},{"name":"viewRoute:development"} ...
```
this is the wrong place to do this in a migration. see [this](https://github.com/jackwrfuller/amazee-lagoon/blob/o2p-authentication/services/keycloak/startup-scripts/00-configure-lagoon.sh#L921-L923) for when additional scopes were added for organization variables

you'll need to do this same thing in a migration, this pr is a good example of that https://github.com/uselagoon/lagoon/pull/3856/files

jackwrfuller replied:
sorry, I'm not quite clear what the action is for me here. Are you saying i should move the "viewRoute:development" and "viewRoute:production" scope migration to `add_org_env_vars`?
